python/sgl_jax/srt/utils/quantization/quantization_utils.py

```diff
@@ -0,0 +1,308 @@
+                                        head_dim=head_dim, layer_num=layer_num,
+                                        mesh=mesh)
+    qwix_rules = parse_qwix_config_to_rules(qwix_config)
+    logits_metadata = LogitsMetadata.from_model_worker_batch(model_worker_batch, mesh)
+    attn_backend.forward_metadata = attn_backend.get_forward_metadata(model_worker_batch)
+    forward_batch = ForwardBatch.init_from_batch(model_worker_batch, mesh, attn_backend)
+    model_input = {
+        "forward_batch": forward_batch,
+        "token_to_kv_pool": token_to_kv_pool,
+        "logits_metadata": logits_metadata,
+    }
+    model = qwix.quantize_model(model, qwix.PtqProvider(qwix_rules),
+                                    **model_input)
+    return model
+
+
+def quantization_config_file_path_to_dict(
+        quantization_config_file_path: str) -> dict:
+    """
+    Converts a quantization config YAML file path to a dictionary.
+
+    The expected format of the quantization config YAML file is as follows:
+    ```yaml
+        qwix:
+            # optional, defaults to False if not specified
+            use_abstract_model: True
+            rules:
+                # NOTE: each entry corresponds to a qwix.QuantizationRule
+                - module_path: '.*attn.*'
+                weight_qtype: 'int8'
+                - module_path: '.*'
+                weight_qtype: 'int8'
+                act_qtype: 'int8'
+    ```
+
+    Args:
+        quantization_config_file_path: the path to the quantization config YAML file
+
+    Returns:
+        a dictionary containing the quantization config
+    """
+    all_entries = os.listdir(QUANTIZATION_CONFIG_PATH)
+    for filename in all_entries:
+        if filename == quantization_config_file_path:
+            path = os.path.join(QUANTIZATION_CONFIG_PATH, filename)
+            with open(path, "r") as f:
+                return yaml.safe_load(f)
+    raise ValueError(
+        f"Could not find quantization config file with name '{quantization_config_file_path}' in 'tpu_inference/models/jax/utils/quantization/configs."
+    )
+
+
+def apply_qwix_quantization(
+        model_config: ModelConfig, model: nnx.Module, mesh: Mesh) -> nnx.Module:
+        from sgl_jax.srt.layers.attention.flashattention_backend import (
+                    FlashAttention,
+)
+        qwix_config_dict = quantization_config_file_path_to_dict(
+                os.path.join(model_config.quantization_config_path)
+        )
+        qwix_config = qwix_config_dict.get("qwix").get("rules")
+        
+        bs = 1
+        max_seq_len = 4096
+        page_size = 1
+        num_tokens = model_config.vocab_size
+        num_attn_heads = model_config.num_attention_heads
+        num_kv_heads = model_config.get_total_num_kv_heads_with_replication(mesh.shape["tensor"])
+        head_dim = model_config.head_dim
+        vocab_size = model_config.vocab_size
+        layer_num = model_config.num_hidden_layers
```

Comment:
I noticed this uses a fixed batch size and context length, and it’s only for decode. This shouldn’t apply to all requests.